### ISSUE TITLE
Use nextflow DSL in a groovy project (IDE completion, documentation, etc.)

### ISSUE THREAD

**Issue by gcornut** [nextflow-io/nextflow#1096]
Hi,

I'm currently testing Nextflow and I was wondering: Is it possible to create a Groovy project importing the Nextflow DSL as a dependency (in a gradle project)? 

It would enable better code re-use and IDE integration while still using the Nextflow DSL

**Comment by pditommaso** [comment]
Technically, should be possible. However, I'm not aware of anybody investigation such extension.

**Comment by gcornut** [comment]
Ok I did some tests with nextflow AST embedded in a groovy script but it pretty complicated to import and instantiate everything. Also I don't think it will really provide IDE integration. 

Another solution might be to create a GDSL file:
http://www.tothenew.com/blog/gdsl-awesomeness-introduction-to-gdsl-in-intellij-idea/

A concrete example with the Jenkins DSL: 
https://st-g.de/2016/08/jenkins-pipeline-autocompletion-in-intellij

What do you think of this? I might do some tests

**Comment by pditommaso** [comment]
Frankly, I've seen before the GDSL file definition. At first looks seems it requires a complete definition of the grammar. in that case, IMO it would make more sense to extend the [groovy grammar](https://github.com/apache/groovy/tree/master/subprojects/parser-antlr4/src/main/java/org/apache/groovy/parser) based on ANTLR4 parser.

**Comment by gcornut** [comment]
Doing some preliminary tests it seems that the GDSL only extends the groovy DSL.

Basically I created an IntelliJ project with a `test.nf` file defined to be a groovy file, then I created a `nextflox.gdsl` file defining the `params` as a property. IntelliJ detected the GDSL file and the use of the `params` property in the `test.nf` was not undefined any more. 

So it seems to work but it would be quite a big investment to define (and maintain) the whole Nextflow DSL as a GDSL file. I could not find any extended documentation on exactly how to do this and don't know much about Groovy DSL.

I think won't invest more into this. But thanks a lot for your help!
